### PR TITLE
Implement `yt-dlp` support and backups for overrides

### LINKED ISSUE #27: Allow users to opt in to downloading the songs instead of streaming
Hi,
As the title says. The music will play after 5 to 10 seconds and that's in my opinion isn't practical. 
So my suggestion is to make your plugin download the music first time the user opens the game page in background. In this case the music will play instantly next times and also this will make the music also play when offline.

### LINKED ISSUE #92: IDEA: Adding downloader for lighten instance and offline playing ?
Hello there, firstly, thanks for your plugin.

I suggest you to lighten bandwidth of instances invidious with a function based on yt dlp based to avoid any appel on instance and overload.

Maybe you’ve already thought about it?

### PR DESCRIPTION
**Pull request by falko17**
This PR adds support for the YouTube downloader `yt-dlp` as an alternative to Invidious instances. Additionally, support for downloading and for creating backups of the overrides have been added. Hence, this should resolve #27, resolve #92, and should provide a stable enough alternative to Invidious that we might consider closing #87 and #89 as well.

## Details
- If enabled, yt-dlp is also used to search for videos. This is done using the `ytsearch10` syntax (i.e., searching for the first 10 matching videos) while restricting results to videos that are not longer than 20 minutes to remove unlikely candidates.
   - The search items are retrieved one by one, so users don't have to wait for all 10 items to load before seeing anything.
- Users can set the plugin up to either download the music or stream it. In the former case, the plugin should also be able to play the music when the device is offline.
- A section has been added to the settings where users can create backups of the list of overrides. When restoring, they can choose from a list of existing backups to go back to.
- A lot of structural changes to the code were made to accommodate these changes (such as introducing an abstract `AudioResolver` class, or moving types into the `types/` directory), I hope none of these are too jarring compared to the old codebase.
- The search text can now be edited while a search is still going on. This was mainly added because yt-dlp searches can be quite slow.
- Additionally, the search text now reflects what's actually being searched (i.e., `appName` followed by "Theme Music" initially).
- A confirmation dialog has been added before deleting all overrides.

## Caveats
- Downloading is currently only supported for yt-dlp. I tried implementing it for Invidious as well by simply downloading the audio URL from the Python backend, but got SSL certificate errors and stopped trying for now[^1]. I've kept the code general, so it should be pretty easy to add that functionality in, assuming the SSL error can be fixed.
- Since decky doesn't seem to have support for external Python libraries, I had to add the binary from the yt-dlp releases directly to the repository. I've also configured it as a "remote binary" in the `package.json`, as is recommended in the decky docs.[^2]
   - Also, even though this binary can be imported as a Python library using `zipimport`, I had to use `subprocess` to execute it as a binary instead. The reason is that when I tried importing it, [there seems to be a problem with accessing `xml.etree` from within decky](https://github.com/SteamDeckHomebrew/decky-loader/issues/322) which I haven't been able to resolve.
- I only speak English and German, so I've only added translations for those languages (and I've fixed some minor details for the German translation in general). 
- I haven't worked with TypeScript nor React before, so I may have implemented some things in unidiomatic ways. Please tell me if you notice anything like that during the review.

[^1]: Currently only one public instance works (the nerdvpn one), so maybe it's just broken on there? But then I'm not sure why streaming it would work.
[^2]: I'm still not exactly sure how this works though, when will the remote binary be downloaded? When testing, I couldn't remove the yt-dlp binary from the `defaults` folder. Will it only be downloaded on production builds?

### CODE REVIEW

**Comment by OMGDuke** [comment]
Hey @falko17 thanks so much for the PR! I've given it a quick look over and everything looks great. I'll find some time to do some testing in the next few days then get it merged in.

Thanks again!

**Comment by eXhumer** [comment]
> I tried implementing it for Invidious as well by simply downloading the audio URL from the Python backend, but got SSL certificate errors and stopped trying for now[1](#user-content-fn-1-c141f79b741bf41cc637111ad0e2bfd5). I've kept the code general, so it should be pretty easy to add that functionality in, assuming the SSL error can be fixed.

How exactly did you try to download the audio in the Python backend? From what I remember, you had to specify an SSL context for SSL connections to work correctly. See [Deckcord](https://github.com/marios8543/Deckcord/blob/804463e1d8127be5feed6042902f04b8a5983c0c/defaults/tab_utils/tab.py#L55-L56) or [decky-loader](https://github.com/SteamDeckHomebrew/decky-loader/blob/43f22329e1b2ca4f707e6c748ea9c013337de8b2/backend/decky_loader/helpers.py#L24) for example.

**Comment by falko17** [comment]
> How exactly did you try to download the audio in the Python backend? From what I remember, you had to specify an SSL context for SSL connections to work correctly. See [Deckcord](https://github.com/marios8543/Deckcord/blob/804463e1d8127be5feed6042902f04b8a5983c0c/defaults/tab_utils/tab.py#L55-L56) or [decky-loader](https://github.com/SteamDeckHomebrew/decky-loader/blob/43f22329e1b2ca4f707e6c748ea9c013337de8b2/backend/decky_loader/helpers.py#L24) for example.

Thanks for the pointers! I just very naively used `urllib.urlretrieve`, but using the same method that your given examples use, it works flawlessly. Hence, as of e0c1f03, Individious downloads are now possible as well.

File: main.py
Comment (eXhumer) [inline main.py:40]: Given the async context of the plugin, it would be better to use [`asyncio.create_subprocess_exec`](https://docs.python.org/3/library/asyncio-subprocess.html#asyncio.create_subprocess_exec).

File: main.py
Comment (eXhumer) [inline main.py:17]: Using `Optional` and importing `typing_extensions` shouldn't be needed here. The backend runs on Python 3.11, we can instead use `subprocess.Popen | None` or `asyncio.subprocess.Process | None` instead instead.

File: main.py
Comment (falko17) [reply main.py]: Agreed, fixed in 49d32b6.

File: main.py
Comment (falko17) [reply main.py]: Right, fixed in 49d32b6.

**Comment by OMGDuke** [comment]
Using yt-dlp with download disabled. Opening a game page with a song that was already set plays fine, but opening search I cant get any songs to load.

When performing a search I'm getting this error

```
Nov 16 16:15:47 steamdeck PluginLoader[1023]: [loader][ERROR]: Method search_yt of plugin Game Theme Music failed with the following exception:
Nov 16 16:15:47 steamdeck PluginLoader[1023]: Traceback (most recent call last):
Nov 16 16:15:47 steamdeck PluginLoader[1023]:   File "decky_loader/loader.py", line 210, in handle_plugin_method_call
Nov 16 16:15:47 steamdeck PluginLoader[1023]:   File "decky_loader/plugin/plugin.py", line 108, in execute_method
Nov 16 16:15:47 steamdeck PluginLoader[1023]:   File "decky_loader/plugin/messages.py", line 35, in wait_for_result
Nov 16 16:15:47 steamdeck PluginLoader[1023]: Exception
```
Any ideas?

**Comment by falko17** [comment]
> Any ideas?

My bad, I didn't test the changes enough when switching to asyncio, and it appears there were some concurrency issues. The critical sections are now guarded by locks.

File: package.json
Comment (eXhumer) [inline package.json:28]: This should be an array of remote binary objects

File: defaults/yt-dlp
Comment (eXhumer) [inline defaults/yt-dlp:1]: This is not needed as the CLI application will automatically fetch, hash verify and package the binary inside the plugin ZIP if a valid remote_binary value is provided

File: main.py
Comment (eXhumer) [inline main.py:46]: Binaries from remote_binary will end up in `f"{decky_plugin.DECKY_PLUGIN_DIR}/bin/yt-dlp"`

File: main.py
Comment (eXhumer) [inline main.py:11]: This is a nitpick, but modern decky interface is provided with `import decky` instead of `import decky_loader`, thought decky-loader provides the exact same import for both really, so it shouldn't matter.

File: main.py
Comment (eXhumer) [inline main.py:31]: Add `await self.yt_process.wait()` to wait for process to terminate.

File: main.py
Comment (eXhumer) [inline main.py:41]: Add `await self.yt_process.wait()` to wait for process to terminate.

File: main.py
Comment (eXhumer) [inline main.py:1]: An overall comment for this file, the `self` argument being used `Plugin` isn't really a real `self`, it is more like the `cls` argument provided by `@classmethod` decorator.

File: package.json
Comment (falko17) [reply package.json]: Ah, I think I got confused by the singular in the name and by [the documentation](https://wiki.deckbrew.xyz/en/plugin-dev/getting-started#package-metadata)—this also explains why it didn't work before. Thanks for pointing it out, the binary is now correctly retrieved!

File: defaults/yt-dlp
Comment (falko17) [reply defaults/yt-dlp:1]: Yes, I removed it.

File: main.py
Comment (falko17) [reply main.py]: Good to know, I changed it to `decky`.

File: main.py
Comment (falko17) [reply main.py:31]: Do we really need to wait for it to terminate on unload, though? It should be enough to just issue the SIGTERM and then just let it terminate in the background, right?

File: main.py
Comment (falko17) [reply main.py:41]: I'm currently using `await self.yt_process.communicate()` instead of `wait()` per [the recommendation here](https://stackoverflow.com/a/78478115). Since a pipe is attached to capture stdout, if yt-dlp tries to output anything without us reading it, this could get us stuck in a deadlock. Or is there another reason to prefer `wait()` here? I don't have a lot of experience with `asyncio` code yet.

File: main.py
Comment (falko17) [reply main.py:1]: I'm a bit confused by this, can you elaborate a bit on it, please? We use this argument to access instance attributes, which wouldn't be possible on a `@classmethod`-like `cls` parameter, so wouldn't this make `self` a more fitting name?

**Comment by falko17** [comment]
I've added another two commits: 37681c5 allows localizations to use placeholders, and e823815 allows downloading all existing overrides.

File: main.py
Comment (eXhumer) [reply main.py:1]: Before api_version 1 of the backend, there was never any instance of an actual Plugin class ever initialized, it looks like this has changed since `"api_version": 1` was introduced. Meaning, you shouldn't need to change anything.

Ref: https://github.com/SteamDeckHomebrew/decky-loader/blob/43f22329e1b2ca4f707e6c748ea9c013337de8b2/backend/decky_loader/plugin/sandboxed_plugin.py#L110-L111

**Comment by eXhumer** [comment]
Sorry for not getting back at this earlier, I was traveling and didn't have reliable internet to be able to do any sort of code reviews.

File: main.py
Comment (eXhumer) [reply main.py:31]: SIGTERM can be ignored by the process, so it is better to either wait for the process to end with some sort of timeout with a later SIGKILL if it doesn't end in time. A process can be left behind in a running state, which we don't be able to kill without a reboot if we don't ensure of the process being ended.

File: main.py
Comment (falko17) [reply main.py:1]: I see, thanks for the explanation!

File: main.py
Comment (falko17) [reply main.py:31]: That's a good point. I changed it in 0e4680c to send SIGKILL if SIGTERM is stuck for more than 5 seconds.

**Comment by falko17** [comment]
> Sorry for not getting back at this earlier, I was traveling and didn't have reliable internet to be able to do any sort of code reviews.

No problem, thanks for your reviews!

**Comment by eXhumer** [comment]
@falko17 Have a look at the PR I opened on your repo. Mostly style changes, translation updates & binary update.

**Comment by libewa** [comment]
We should maybe consider having a slider to adjust the minimum length.

**Comment by falko17** [comment]
> We should maybe consider having a slider to adjust the minimum length.

That sounds like a good idea, but I think I'd do that in a separate issue and PR after this one has been merged.

**Comment by BrentTechDesigns** [comment]
@falko17 Do you have a Ko-Fi or somewhere that we could support your efforts?  I really appreciate the work you’ve put in to resurrect this plugin.

**Review by OMGDuke — APPROVED** [review]
Approved, thanks for all your hard work @falko17 and thanks for the reviews @eXhumer

**Comment by falko17** [comment]
> @falko17 Do you have a Ko-Fi or somewhere that we could support your efforts?

That's very kind, thanks @BrentTechDesigns! [I do now](https://ko-fi.com/falko17), but it'll probably be at least a few weeks before I implement another feature for this plugin.

> Approved

Great, thanks, and thank you to @eXhumer from me as well for the helpful comments.

**Comment by libewa** [comment]
@OMGDuke hows the submission going? i dont see the update.

**Comment by OMGDuke** [comment]
> @OMGDuke hows the submission going? i dont see the update.

https://github.com/SteamDeckHomebrew/decky-plugin-database/pull/730

Awaiting review from the Decky team. No ETA unfortunately. They're all volunteers.

**Pull request merged**
